File: src/main/java/cn/wjybxx/dson/DsonBinary.java
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import javax.annotation.Nonnull;
import java.util.Arrays;

/**
 * 你通常不应该修改data中的数据。
 * 该类难以实现不可变对象，虽然我们可以封装为ByteArray，
 * 但许多接口都是基于byte[]的，封装会导致难以与其它接口协作。
 *
 * @author wjybxx
 * date - 2023/4/19
 */
public class DsonBinary extends DsonValue {

    private final int type;
    private final byte[] data;

    public DsonBinary(byte[] data) {
        this(0, data);
    }

    public DsonBinary(int type, byte[] data) {
        Dsons.checkSubType(type);
        Dsons.checkBinaryLength(data.length); // 顺带NPE
        this.type = type;
        this.data = data;
    }

    /** 默认采取拷贝的方式，保证安全性 */
    public DsonBinary(DsonBinary src) {
        this.type = src.type;
        this.data = src.data.clone();
    }

    @Nonnull
    @Override
    public DsonType getDsonType() {
        return DsonType.BINARY;
    }

    public int getType() {
        return type;
    }

    /** 最好不要持有data的引用，否则可能由于共享数据产生错误 */
    public byte[] getData() {
        return data;
    }

    /** 创建一个拷贝 */
    public DsonBinary copy() {
        return new DsonBinary(type, data.clone());
    }

    //region equals

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        DsonBinary that = (DsonBinary) o;

        if (type != that.type) return false;
        return Arrays.equals(data, that.data);
    }

    @Override
    public int hashCode() {
        int result = type;
        result = 31 * result + Arrays.hashCode(data);
        return result;
    }

    // endregion

    @Override
    public String toString() {
        return "DsonBinary{" +
                "type=" + type +
                ", data=" + Arrays.toString(data) +
                '}';
    }

}

File: csharp/Dson/DsonBinary.cs
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using Dson.IO;

namespace Dson;

public class DsonBinary : DsonValue, IEquatable<DsonBinary>
{
    private readonly int _type;
    private readonly byte[] _data;

    public DsonBinary(int type, byte[] data) {
        if (data == null) throw new ArgumentNullException(nameof(data));
        Dsons.CheckSubType(type);
        Dsons.CheckBinaryLength(data.Length);
        _type = type;
        _data = data;
    }

    public DsonBinary(int type, DsonChunk chunk) {
        if (chunk == null) throw new ArgumentNullException(nameof(chunk));
        Dsons.CheckSubType(type);
        Dsons.CheckBinaryLength(chunk.Length);
        _type = type;
        _data = chunk.Payload();
    }

    public DsonBinary(DsonBinary src) {
        this._type = src._type;
        this._data = (byte[])src._data.Clone();
    }

    public override DsonType DsonType => DsonType.BINARY;
    public int Type => _type;

    /// <summary>
    /// 不宜修改返回的数据
    /// </summary>
    public byte[] Data => _data;

    /// <summary>
    /// 创建一个副本
    /// </summary>
    /// <returns></returns>
    public DsonBinary Copy() {
        return new DsonBinary(this);
    }

    #region equals

    public bool Equals(DsonBinary? other) {
        if (ReferenceEquals(null, other)) return false;
        if (ReferenceEquals(this, other)) return true;
        return _type == other._type && _data.SequenceEqual(other._data); // 比较数组元素需要使用SequenceEqual
    }

    public override bool Equals(object? obj) {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != this.GetType()) return false;
        return Equals((DsonBinary)obj);
    }

    public override int GetHashCode() {
        return HashCode.Combine(_type, _data);
    }

    public static bool operator ==(DsonBinary? left, DsonBinary? right) {
        return Equals(left, right);
    }

    public static bool operator !=(DsonBinary? left, DsonBinary? right) {
        return !Equals(left, right);
    }

    #endregion

    public override string ToString() {
        return $"{nameof(DsonType)}: {DsonType}, {nameof(_type)}: {_type}, {nameof(_data)}: {Convert.ToHexString(_data)}";
    }
}